google/cloud/pubsub_v1/open_telemetry/publish_message_wrapper.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2017, Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+from datetime import datetime
+import warnings
+from typing import Optional
+
+from opentelemetry import trace
+from opentelemetry.trace.propagation import set_span_in_context
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+
+from google.pubsub_v1 import types as gapic_types
+from google.cloud.pubsub_v1.open_telemetry.context_propagation import (
+    OpenTelemetryContextSetter,
+)
+
+
+class PublishMessageWrapper:
+    _OPEN_TELEMETRY_TRACER_NAME: str = "google.cloud.pubsub_v1.publisher"
+    _OPEN_TELEMETRY_MESSAGING_SYSTEM: str = "gcp_pubsub"
+    _OPEN_TELEMETRY_PUBLISHER_BATCHING = "publisher batching"
+
+    _PUBLISH_START_EVENT: str = "publish start"
+    _PUBLISH_FLOW_CONTROL: str = "publisher flow control"
+
+    def __init__(self, message: gapic_types.PubsubMessage):
+        self._message: gapic_types.PubsubMessage = message
+
+    @property
+    def message(self):
+        return self._message
+
+    @message.setter  # type: ignore[no-redef]  # resetting message value is intentional here
+    def message(self, message: gapic_types.PubsubMessage):
+        self._message = message
+
+    @property
+    def create_span(self):
+        return self._create_span
+
+    def __eq__(self, other):  # pragma: NO COVER
+        """Used for pytest asserts to compare two PublishMessageWrapper objects with the same message"""
+        if isinstance(self, other.__class__):
+            return self.message == other.message
+        return False
+
+    def start_create_span(self, topic: str, ordering_key: str) -> None:
+        tracer = trace.get_tracer(self._OPEN_TELEMETRY_TRACER_NAME)
+        topic_short_name = topic.split("/")[3]
```

Comment:
We could do this as a precautionary measure, but the statement as is matches format and length of this topic specified in the public documentation: https://cloud.google.com/pubsub/docs/publish-receive-messages-client-library#publish_messages

Reply:
In other languages, it is possible that if the user creates a fake of the pubsub service, they might return a topic not in the format that's specified in the public documentation. This might be worth guarding around.

Reply:
Adding an assert on the length of the topic. 

I checked what happens if the  users tries to publish messages in a format that's not topic=`projects/projectID/topics/topicID` in the current state of the library(i.e without Open Telemetry). It fails as expected:

```
Published messages with error handler to mukund-topic.
exception calling callback for <Future at 0x7f9b5bb51fd0 state=finished raised InvalidArgument>
Traceback (most recent call last):
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 340, in _invoke_callbacks
    callback(self)
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/publish.py", line 21, in callback
    print(publish_future.result(timeout=60))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/google/cloud/pubsub_v1/publisher/futures.py", line 66, in result
    return super().result(timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/google/cloud/pubsub_v1/publisher/_batch/thread.py", line 274, in _commit
    response = self._client._gapic_publish(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/google/cloud/pubsub_v1/publisher/client.py", line 267, in _gapic_publish
    return super().publish(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/google/pubsub_v1/services/publisher/client.py", line 1067, in publish
    response = rpc(
               ^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    return wrapped_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 293, in retry_wrapped_func
    return retry_target(
           ^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 153, in retry_target
    _retry_error_helper(
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/retry/retry_base.py", line 212, in _retry_error_helper
    raise final_exc from source_exc
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 144, in retry_target
    result = target()
             ^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/timeout.py", line 120, in func_with_timeout
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/mukundananthu/debug/python-pubsub/.venv/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 78, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InvalidArgument: 400 Invalid resource name given (name=mukund-topic). Refer to https://cloud.google.com/pubsub/docs/pubsub-basics#resource_names for more information. [detail: "Could not parse topic \'mukund-topic\'"
]
```

This is because the topic supplied is bubbled down all the way to gapic publish, which on attempting to publish to the backend with a topic specified in any other format, would get the above error.

Reply:
It would be incorrect implementation of the fake, correct. It's something I had dealt with in Go (where some of our tests allows creating a topic without a valid project ID). Anyways I don't feel strongly about this so I'm ok with leaving as is.

Reply:
Mentioning in case this got missed - I have added an assert to guard against this case `Adding an assert on the length of the topic.`.